### PR TITLE
Update NSView with setShadow:/shadow methods

### PR DESCRIPTION
**Pull request by gcasa**
Please see the bug here: https://github.com/gnustep/libs-gui/issues/217

### CODE REVIEW

**Comment by gcasa** [comment]
I think that this might give some clue as to how to do this in the backend.  I am not sure that shadows necessarily need layers to work.
https://stackoverflow.com/questions/30630280/how-do-i-create-a-circular-shadow-effect-with-cairo

**Comment by gcasa** [comment]
Also of interest MIGHT be... https://github.com/rcairo/rcairo/blob/master/lib/cairo/context/blur.rb

I know these examples are in different languages, but they are both using cairo, so I wanted to note them here.

**Comment by gcasa** [comment]
It seems as though the keys being used are these...
```
[heron@nomad Base.lproj] % strings MainMenu.nib | grep -i shadow                                     [0]
NSShadowHoriz
IBNSShadowedSymbolConfiguration
NSShadowVert
NSViewShadow
NSShadowBlurRadius
NSShadowColor
NSShadow_Test
NSShadow_Test
Quit NSShadow_Test
Quit NSShadow_Test
About NSShadow_Test
About NSShadow_Test
NSShadow_Test Help
NSShadow_Test Help
Hide NSShadow_Test
Hide NSShadow_Test
NSShadow_Test
NSShadow_Test
NSShadow
```
The new NIBArchive format is proving to be a menace at analyzing nib files for keys... but since we can read xib files it is not a problem, but it is a format we might want to consider looking at in the future. GC

**Comment by gcasa** [comment]
Yet more info about blur... https://discourse.gnome.org/t/cairo-gaussian-blur-for-shadows/2174

File: Source/NSView.m [hunk 1]
```diff
@@ -4634,7 +4646,7 @@ - (id) initWithCoder: (NSCoder*)aDecoder
   NSEnumerator *e;
   NSView	*sub;
   NSArray	*subs;
-
+
```
Comment (fredkiefer) [inline Source/NSView.m:4649]: Here is some leftover space

Comment (gcasa) [reply Source/NSView.m]: Should be fixed.

**Comment by gcasa** [comment]
Okay, so now I have got everything decoding for the NSShadow from the XIB properly.   I am going to work on the backend now.   Some changes might be needed in libs-gui, but I doubt it.

**Comment by fredkiefer** [comment]
Let me try to explain this once more:
- You should start off by implementing direct drawing of the shadow. For this you will need to add a setShadow: method on the GSGraphicsContext class to be called from `[NSShadow set]`. After that you just need to implement that method or rather the actual handling of the shadow while drawing in the Cairo graphics context in the backend.
- Implementing the shadow drawing for a view requires more changes in gui. When drawing a view we always clip to the frame of the view. This means, when drawing the shadow from the view drawing code, it would never be visible. Most likely we need to draw it from the super view context somewhere in the whole view hierarchy drawing code. But this is only part of the problem. You also need to figure out what parts of the screen need to be invalidated when a view with a shadow requires a redraw. Please take your time here. A complete day just trying to understand the existing code would be well spend. Then figure out what extensions will be required and before implementing them, let us discuss your approach.

I would suggest to merge this PR here separately and not wait until you have the complete implementation in place.

**Comment by gcasa** [comment]
Okay.  Thank you for the great input.  I will change this PR to ready and if you find it suitable, then please review.

**Comment by gcasa** [comment]
I will open another PR for the shadow work in NSGraphicsContext.

**Comment by gcasa** [comment]
@fredkiefer please approve if it's okay to merge.  I am in the process of looking at what is involved on the backend.

**Comment by gcasa** [comment]
https://github.com/gnustep/libs-gui/pull/232 PR for set method on libs-gui

File: Source/GSXib5KeyedUnarchiver.m [hunk 2]
```diff
@@ -445,6 +448,8 @@ + (void) initialize
                @"decodeSecondAttribute:", @"NSSecondAttribute",
                @"decodeRelation:", @"NSRelation",
                @"decodeTransitionStyle:", @"NSTransitionStyle",
+	    @"decodeShadowOffsetHoriz:", @"NSShadowHoriz",
```
Comment (fredkiefer) [inline Source/GSXib5KeyedUnarchiver.m:451]: Looks like the indentation is incorrect

File: Source/GSXib5KeyedUnarchiver.m [hunk 3]
```diff
@@ -2984,6 +2989,20 @@ - (id) decodeTransitionStyle: (GSXibElement *)element
   return num;  
 }
 
+- (id) decodeShadowOffsetHoriz: (GSXibElement *)element
+{
+  NSSize size = [self decodeSizeForKey: @"offset"];
+  NSNumber *num = [NSNumber numberWithFloat: size.width];
```
Comment (fredkiefer) [inline Source/GSXib5KeyedUnarchiver.m:2995]: This local variable is not needed.

File: Source/NSView.m [hunk 4]
```diff
@@ -4737,12 +4749,19 @@ - (id) initWithCoder: (NSCoder*)aDecoder
 	  [self didAddSubview: sub];
 	}
 
+      // Decode the shadow...
+      if ([aDecoder containsValueForKey: @"NSViewShadow"])
+	{
+	  _shadow = [aDecoder decodeObjectForKey: @"NSViewShadow"];
```
Comment (fredkiefer) [inline Source/NSView.m:4755]: Here a RETAIN would be required.

File: Source/GSXib5KeyedUnarchiver.m (hunk 3 above)
Comment (gcasa) [reply Source/GSXib5KeyedUnarchiver.m]: Removed

File: Source/GSXib5KeyedUnarchiver.m (hunk 2 above)
Comment (gcasa) [reply Source/GSXib5KeyedUnarchiver.m]: Corrected

File: Source/NSView.m [hunk 5]
```diff
@@ -4798,6 +4817,12 @@ - (id) initWithCoder: (NSCoder*)aDecoder
 	  [self didAddSubview: sub];
 	}
       RELEASE(subs);
+
+      // Decode the shadow if this is version 2 or greater...
+      if (version >= 2)
+	{
+	  _shadow = [aDecoder decodeObject];
```
Comment (fredkiefer) [inline Source/NSView.m:4824]: Again you probably should retain the decoded object as you release it in the dealloc method.

File: Source/NSView.m (hunk 4 above)
Comment (gcasa) [reply Source/NSView.m]: Done

File: Source/NSView.m (hunk 5 above)
Comment (gcasa) [reply Source/NSView.m]: Done.

**Review by fredkiefer — CHANGES_REQUESTED** [review]
Looks mostly OK, but the RETAIN issue needs to be resolved. You could also use the setter instead.

**Review by fredkiefer — APPROVED** [review]

**Comment by gcasa** [comment]
> Looks mostly OK, but the RETAIN issue needs to be resolved. You could also use the setter instead.

I added both retains, I think using the setter is cleaner.  I am going to go ahead and make that change.  Give me a moment.

**Comment by gcasa** [comment]
I changed the code to use the setter, thanks. :). Merging.

**Comment by gcasa** [comment]
Forgot to add the NSView.m file to the commit... I will correct it on master. Sorry.

**Comment by gcasa** [comment]
Oh... it got them.  :). Nevermind.

**Pull request merged**